File: string-algorithms/suffix_array.go
/*
 * Suffix Array and LCP Array Construction in Go
 * =============================================
 *
 * Algorithms implemented:
 * 1. Prefix doubling O(n log² n) for suffix array construction
 * 2. Kasai's algorithm O(n) for LCP array computation
 *
 * Applications:
 * - Pattern matching in O(m log n) time
 * - Finding longest repeated substring
 * - Counting distinct substrings
 * - Longest common substring between two strings
 *
 * Usage: go run suffix_array.go
 */

package main

import (
	"fmt"
	"sort"
	"strings"
	"time"
)

// Suffix represents a suffix with its index and rank pair
type Suffix struct {
	Index int
	Rank  [2]int
}

// SuffixArray provides methods for suffix array construction and operations
type SuffixArray struct {
	text string
	sa   []int
	lcp  []int
}

// NewSuffixArray creates a new suffix array from text
func NewSuffixArray(text string) *SuffixArray {
	sa := &SuffixArray{
		text: text,
	}
	sa.Build()
	sa.BuildLCP()
	return sa
}

// Build constructs the suffix array using prefix doubling algorithm
// Time: O(n log² n)
// Space: O(n)
func (sa *SuffixArray) Build() {
	n := len(sa.text)
	sa.sa = make([]int, n)

	if n == 0 {
		return
	}

	// Create suffix structures
	suffixes := make([]Suffix, n)
	rank := make([]int, n)

	// Initialize ranks with character values
	for i := 0; i < n; i++ {
		suffixes[i].Index = i
		suffixes[i].Rank[0] = int(sa.text[i])
		if i+1 < n {
			suffixes[i].Rank[1] = int(sa.text[i+1])
		} else {
			suffixes[i].Rank[1] = -1
		}
	}

	// Sort by first two characters
	sort.Slice(suffixes, func(i, j int) bool {
		if suffixes[i].Rank[0] != suffixes[j].Rank[0] {
			return suffixes[i].Rank[0] < suffixes[j].Rank[0]
		}
		return suffixes[i].Rank[1] < suffixes[j].Rank[1]
	})

	// Build initial rank array
	for i := 0; i < n; i++ {
		rank[suffixes[i].Index] = i
	}

	// Prefix doubling
	for k := 4; k < 2*n; k *= 2 {
		// Update ranks for sorting
		for i := 0; i < n; i++ {
			curr := suffixes[i].Index
			suffixes[i].Rank[0] = rank[curr]
			if curr+k/2 < n {
				suffixes[i].Rank[1] = rank[curr+k/2]
			} else {
				suffixes[i].Rank[1] = -1
			}
		}

		// Sort suffixes
		sort.Slice(suffixes, func(i, j int) bool {
			if suffixes[i].Rank[0] != suffixes[j].Rank[0] {
				return suffixes[i].Rank[0] < suffixes[j].Rank[0]
			}
			return suffixes[i].Rank[1] < suffixes[j].Rank[1]
		})

		// Update ranks
		tempRank := make([]int, n)
		tempRank[suffixes[0].Index] = 0

		for i := 1; i < n; i++ {
			prev := suffixes[i-1]
			curr := suffixes[i]

			// Same rank if both pairs are equal
			if curr.Rank[0] == prev.Rank[0] && curr.Rank[1] == prev.Rank[1] {
				tempRank[curr.Index] = tempRank[prev.Index]
			} else {
				tempRank[curr.Index] = tempRank[prev.Index] + 1
			}
		}

		rank = tempRank
	}

	// Build suffix array from sorted suffixes
	for i := 0; i < n; i++ {
		sa.sa[i] = suffixes[i].Index
	}
}

// BuildLCP builds the LCP array using Kasai's algorithm
// Time: O(n)
// Space: O(n)
//
// LCP[i] = length of longest common prefix between
//          suffix[SA[i]] and suffix[SA[i-1]]
func (sa *SuffixArray) BuildLCP() {
	n := len(sa.text)
	sa.lcp = make([]int, n)

	if n == 0 {
		return
	}

	// Compute inverse suffix array (rank)
	rank := make([]int, n)
	for i := 0; i < n; i++ {
		rank[sa.sa[i]] = i
	}

	h := 0 // Height of LCP

	for i := 0; i < n; i++ {
		if rank[i] > 0 {
			j := sa.sa[rank[i]-1]

			// Compute LCP
			for i+h < n && j+h < n && sa.text[i+h] == sa.text[j+h] {
				h++
			}

			sa.lcp[rank[i]] = h

			// Decrease h for next iteration
			if h > 0 {
				h--
			}
		}
	}
}

// PatternSearch searches for pattern using binary search on suffix array
// Time: O(m log n)
// Returns: Slice of starting positions
func (sa *SuffixArray) PatternSearch(pattern string) []int {
	n := len(sa.text)
	m := len(pattern)
	matches := []int{}

	if m == 0 || n == 0 {
		return matches
	}

	// Binary search for lower bound
	lower := sort.Search(n, func(i int) bool {
		suffix := sa.text[sa.sa[i]:]
		return suffix >= pattern
	})

	// Binary search for upper bound
	upper := sort.Search(n, func(i int) bool {
		suffix := sa.text[sa.sa[i]:]
		if len(suffix) < m {
			return suffix > pattern[:len(suffix)]
		}
		return suffix[:m] > pattern
	})

	// Collect all matches
	for i := lower; i < upper; i++ {
		if strings.HasPrefix(sa.text[sa.sa[i]:], pattern) {
			matches = append(matches, sa.sa[i])
		}
	}

	return matches
}

// LongestRepeatedSubstring finds the longest repeated substring
// Time: O(n)
func (sa *SuffixArray) LongestRepeatedSubstring() string {
	if len(sa.lcp) == 0 {
		return ""
	}

	maxLCP := 0
	maxIdx := 0

	// Find maximum LCP value
	for i := 1; i < len(sa.lcp); i++ {
		if sa.lcp[i] > maxLCP {
			maxLCP = sa.lcp[i]
			maxIdx = i
		}
	}

	if maxLCP == 0 {
		return ""
	}

	return sa.text[sa.sa[maxIdx] : sa.sa[maxIdx]+maxLCP]
}

// CountDistinctSubstrings counts the number of distinct substrings
// Time: O(n)
// Formula: n*(n+1)/2 - sum(LCP)
func (sa *SuffixArray) CountDistinctSubstrings() int64 {
	n := int64(len(sa.text))
	total := n * (n + 1) / 2

	var duplicates int64
	for i := 1; i < len(sa.lcp); i++ {
		duplicates += int64(sa.lcp[i])
	}

	return total - duplicates
}

// Visualize prints the suffix array and LCP array
func (sa *SuffixArray) Visualize() {
	fmt.Printf("\nSuffix Array Visualization:\n")
	fmt.Printf("Text: '%s'\n", sa.text)
	fmt.Printf("Length: %d\n\n", len(sa.text))

	fmt.Printf("%-4s | %-6s | %-6s | Suffix\n", "i", "SA[i]", "LCP[i]")
	fmt.Println(strings.Repeat("-", 70))

	for i := 0; i < len(sa.sa); i++ {
		suffix := sa.text[sa.sa[i]:]
		if len(suffix) > 40 {
			suffix = suffix[:37] + "..."
		}
		fmt.Printf("%-4d | %-6d | %-6d | %s\n", i, sa.sa[i], sa.lcp[i], suffix)
	}
	fmt.Println()
}

// LongestCommonSubstring finds the longest common substring between two strings
// Time: O(n + m)
func LongestCommonSubstring(text1, text2 string) string {
	// Concatenate with separator
	separator := "#"
	combined := text1 + separator + text2
	len1 := len(text1)

	// Build suffix array
	sa := NewSuffixArray(combined)

	// Find max LCP where adjacent suffixes are from different strings
	maxLCP := 0
	maxPos := 0

	for i := 1; i < len(sa.sa); i++ {
		pos1 := sa.sa[i-1]
		pos2 := sa.sa[i]

		// One before separator, one after
		if (pos1 < len1) != (pos2 < len1) {
			if sa.lcp[i] > maxLCP {
				maxLCP = sa.lcp[i]
				maxPos = sa.sa[i]
			}
		}
	}

	if maxLCP == 0 {
		return ""
	}

	return combined[maxPos : maxPos+maxLCP]
}

// Benchmark measures suffix array construction time
func BenchmarkConstruction(text string) time.Duration {
	start := time.Now()
	NewSuffixArray(text)
	return time.Since(start)
}

// =========================================================================
// EXAMPLES AND TESTING
// =========================================================================

func exampleBasicConstruction() {
	fmt.Println(strings.Repeat("=", 70))
	fmt.Println("EXAMPLE 1: Basic Suffix Array Construction")
	fmt.Println(strings.Repeat("=", 70))

	text := "banana"
	sa := NewSuffixArray(text)
	sa.Visualize()
}

func examplePatternSearch() {
	fmt.Println(strings.Repeat("=", 70))
	fmt.Println("EXAMPLE 2: Pattern Searching")
	fmt.Println(strings.Repeat("=", 70))

	text := "the quick brown fox jumps over the lazy dog"
	pattern := "the"

	sa := NewSuffixArray(text)
	matches := sa.PatternSearch(pattern)

	fmt.Printf("Text: '%s'\n", text)
	fmt.Printf("Pattern: '%s'\n", pattern)
	fmt.Printf("Matches at positions: %v\n\n", matches)

	for _, pos := range matches {
		fmt.Printf("  Position %d: '%s'\n", pos, text[pos:pos+len(pattern)])
	}
	fmt.Println()
}

func exampleLongestRepeatedSubstring() {
	fmt.Println(strings.Repeat("=", 70))
	fmt.Println("EXAMPLE 3: Longest Repeated Substring")
	fmt.Println(strings.Repeat("=", 70))

	text := "abracadabra"
	sa := NewSuffixArray(text)
	lrs := sa.LongestRepeatedSubstring()

	fmt.Printf("Text: '%s'\n", text)
	fmt.Printf("Longest repeated substring: '%s'\n\n", lrs)
}

func exampleCountDistinct() {
	fmt.Println(strings.Repeat("=", 70))
	fmt.Println("EXAMPLE 4: Count Distinct Substrings")
	fmt.Println(strings.Repeat("=", 70))

	text := "abab"
	sa := NewSuffixArray(text)
	count := sa.CountDistinctSubstrings()

	fmt.Printf("Text: '%s'\n", text)
	fmt.Printf("Number of distinct substrings: %d\n\n", count)
}

func exampleLongestCommonSubstring() {
	fmt.Println(strings.Repeat("=", 70))
	fmt.Println("EXAMPLE 5: Longest Common Substring")
	fmt.Println(strings.Repeat("=", 70))

	text1 := "algorithms"
	text2 := "altruistic"
	lcs := LongestCommonSubstring(text1, text2)

	fmt.Printf("Text 1: '%s'\n", text1)
	fmt.Printf("Text 2: '%s'\n", text2)
	fmt.Printf("Longest common substring: '%s'\n\n", lcs)
}

func exampleBenchmark() {
	fmt.Println(strings.Repeat("=", 70))
	fmt.Println("EXAMPLE 6: Performance Benchmarking")
	fmt.Println(strings.Repeat("=", 70))

	// Small text
	smallText := strings.Repeat("banana", 10)
	smallTime := BenchmarkConstruction(smallText)
	fmt.Printf("Small text (length %d): %v\n", len(smallText), smallTime)

	// Medium text
	mediumText := strings.Repeat("abracadabra", 100)
	mediumTime := BenchmarkConstruction(mediumText)
	fmt.Printf("Medium text (length %d): %v\n\n", len(mediumText), mediumTime)
}

func main() {
	fmt.Println(strings.Repeat("=", 70))
	fmt.Println("SUFFIX ARRAY AND LCP ARRAY IN GO")
	fmt.Println(strings.Repeat("=", 70))
	fmt.Println()

	exampleBasicConstruction()
	examplePatternSearch()
	exampleLongestRepeatedSubstring()
	exampleCountDistinct()
	exampleLongestCommonSubstring()
	exampleBenchmark()

	fmt.Println(strings.Repeat("=", 70))
	fmt.Println("All examples completed successfully!")
	fmt.Println(strings.Repeat("=", 70))
}

File: string-algorithms/suffix_array.rs
/*
 * Suffix Array and LCP Array Construction in Rust
 * ===============================================
 *
 * Algorithms implemented:
 * 1. Prefix doubling O(n log² n) for suffix array construction
 * 2. Kasai's algorithm O(n) for LCP array computation
 *
 * Applications:
 * - Pattern matching in O(m log n) time
 * - Finding longest repeated substring
 * - Counting distinct substrings
 * - Longest common substring between two strings
 *
 * Compilation: rustc -O suffix_array.rs
 * Usage: ./suffix_array
 */

use std::time::Instant;

/// Suffix structure for sorting
#[derive(Debug, Clone, Copy)]
struct Suffix {
    index: usize,
    rank: [i32; 2],
}

/// Suffix Array with LCP array
pub struct SuffixArray {
    text: String,
    sa: Vec<usize>,
    lcp: Vec<usize>,
}

impl SuffixArray {
    /// Create a new suffix array from text
    pub fn new(text: &str) -> Self {
        let mut sa = SuffixArray {
            text: text.to_string(),
            sa: Vec::new(),
            lcp: Vec::new(),
        };
        sa.build();
        sa.build_lcp();
        sa
    }

    /// Build suffix array using prefix doubling algorithm
    /// Time: O(n log² n)
    /// Space: O(n)
    fn build(&mut self) {
        let n = self.text.len();
        if n == 0 {
            return;
        }

        let text_bytes = self.text.as_bytes();

        // Create suffix structures
        let mut suffixes: Vec<Suffix> = (0..n)
            .map(|i| Suffix {
                index: i,
                rank: [
                    text_bytes[i] as i32,
                    if i + 1 < n { text_bytes[i + 1] as i32 } else { -1 },
                ],
            })
            .collect();

        // Sort by first two characters
        suffixes.sort_by(|a, b| {
            if a.rank[0] != b.rank[0] {
                a.rank[0].cmp(&b.rank[0])
            } else {
                a.rank[1].cmp(&b.rank[1])
            }
        });

        // Build initial rank array
        let mut rank = vec![0; n];
        for (i, suffix) in suffixes.iter().enumerate() {
            rank[suffix.index] = i as i32;
        }

        // Prefix doubling
        let mut k = 4;
        while k < 2 * n {
            // Update ranks for sorting
            for i in 0..n {
                let curr = suffixes[i].index;
                suffixes[i].rank[0] = rank[curr];
                suffixes[i].rank[1] = if curr + k / 2 < n {
                    rank[curr + k / 2]
                } else {
                    -1
                };
            }

            // Sort suffixes
            suffixes.sort_by(|a, b| {
                if a.rank[0] != b.rank[0] {
                    a.rank[0].cmp(&b.rank[0])
                } else {
                    a.rank[1].cmp(&b.rank[1])
                }
            });

            // Update ranks
            let mut temp_rank = vec![0; n];
            temp_rank[suffixes[0].index] = 0;

            for i in 1..n {
                let prev = suffixes[i - 1];
                let curr = suffixes[i];

                // Same rank if both pairs are equal
                temp_rank[curr.index] = if curr.rank[0] == prev.rank[0]
                    && curr.rank[1] == prev.rank[1]
                {
                    temp_rank[prev.index]
                } else {
                    temp_rank[prev.index] + 1
                };
            }

            rank = temp_rank;
            k *= 2;
        }

        // Build suffix array from sorted suffixes
        self.sa = suffixes.iter().map(|s| s.index).collect();
    }

    /// Build LCP array using Kasai's algorithm
    /// Time: O(n)
    /// Space: O(n)
    ///
    /// LCP[i] = length of longest common prefix between
    ///          suffix[SA[i]] and suffix[SA[i-1]]
    fn build_lcp(&mut self) {
        let n = self.text.len();
        self.lcp = vec![0; n];

        if n == 0 {
            return;
        }

        let text_bytes = self.text.as_bytes();

        // Compute inverse suffix array (rank)
        let mut rank = vec![0; n];
        for (i, &sa_i) in self.sa.iter().enumerate() {
            rank[sa_i] = i;
        }

        let mut h = 0; // Height of LCP

        for i in 0..n {
            if rank[i] > 0 {
                let j = self.sa[rank[i] - 1];

                // Compute LCP
                while i + h < n && j + h < n && text_bytes[i + h] == text_bytes[j + h] {
                    h += 1;
                }

                self.lcp[rank[i]] = h;

                // Decrease h for next iteration
                if h > 0 {
                    h -= 1;
                }
            }
        }
    }

    /// Pattern search using binary search on suffix array
    /// Time: O(m log n)
    pub fn pattern_search(&self, pattern: &str) -> Vec<usize> {
        let n = self.text.len();
        let m = pattern.len();
        let mut matches = Vec::new();

        if m == 0 || n == 0 {
            return matches;
        }

        // Binary search for lower bound
        let lower = self.sa.partition_point(|&idx| {
            let suffix = &self.text[idx..];
            suffix < pattern
        });

        // Binary search for upper bound
        let upper = self.sa.partition_point(|&idx| {
            let suffix = &self.text[idx..];
            if suffix.len() < m {
                suffix <= &pattern[..suffix.len()]
            } else {
                &suffix[..m] <= pattern
            }
        });

        // Collect all matches
        for i in lower..upper {
            if self.text[self.sa[i]..].starts_with(pattern) {
                matches.push(self.sa[i]);
            }
        }

        matches
    }

    /// Find longest repeated substring
    /// Time: O(n)
    pub fn longest_repeated_substring(&self) -> &str {
        if self.lcp.is_empty() {
            return "";
        }

        let mut max_lcp = 0;
        let mut max_idx = 0;

        // Find maximum LCP value
        for (i, &lcp_val) in self.lcp.iter().enumerate().skip(1) {
            if lcp_val > max_lcp {
                max_lcp = lcp_val;
                max_idx = i;
            }
        }

        if max_lcp == 0 {
            return "";
        }

        &self.text[self.sa[max_idx]..self.sa[max_idx] + max_lcp]
    }

    /// Count distinct substrings
    /// Time: O(n)
    /// Formula: n*(n+1)/2 - sum(LCP)
    pub fn count_distinct_substrings(&self) -> usize {
        let n = self.text.len();
        let total = n * (n + 1) / 2;
        let duplicates: usize = self.lcp.iter().sum();
        total - duplicates
    }

    /// Visualize suffix array and LCP array
    pub fn visualize(&self) {
        println!("\nSuffix Array Visualization:");
        println!("Text: '{}'", self.text);
        println!("Length: {}\n", self.text.len());

        println!("{:<4} | {:<6} | {:<6} | Suffix", "i", "SA[i]", "LCP[i]");
        println!("{}", "-".repeat(70));

        for (i, &sa_i) in self.sa.iter().enumerate() {
            let suffix = &self.text[sa_i..];
            let suffix_display = if suffix.len() > 40 {
                format!("{}...", &suffix[..37])
            } else {
                suffix.to_string()
            };
            println!("{:<4} | {:<6} | {:<6} | {}", i, sa_i, self.lcp[i], suffix_display);
        }
        println!();
    }
}

/// Find longest common substring between two strings
/// Time: O(n + m)
pub fn longest_common_substring(text1: &str, text2: &str) -> String {
    // Concatenate with separator
    let separator = "#";
    let combined = format!("{}{}{}", text1, separator, text2);
    let len1 = text1.len();

    // Build suffix array
    let sa = SuffixArray::new(&combined);

    // Find max LCP where adjacent suffixes are from different strings
    let mut max_lcp = 0;
    let mut max_pos = 0;

    for i in 1..sa.sa.len() {
        let pos1 = sa.sa[i - 1];
        let pos2 = sa.sa[i];

        // One before separator, one after
        if (pos1 < len1) != (pos2 < len1) {
            if sa.lcp[i] > max_lcp {
                max_lcp = sa.lcp[i];
                max_pos = sa.sa[i];
            }
        }
    }

    if max_lcp == 0 {
        return String::new();
    }

    combined[max_pos..max_pos + max_lcp].to_string()
}

/// Benchmark suffix array construction
fn benchmark_construction(text: &str) -> std::time::Duration {
    let start = Instant::now();
    SuffixArray::new(text);
    start.elapsed()
}

// =========================================================================
// EXAMPLES AND TESTING
// =========================================================================

fn example_basic_construction() {
    println!("{}", "=".repeat(70));
    println!("EXAMPLE 1: Basic Suffix Array Construction");
    println!("{}", "=".repeat(70));

    let text = "banana";
    let sa = SuffixArray::new(text);
    sa.visualize();
}

fn example_pattern_search() {
    println!("{}", "=".repeat(70));
    println!("EXAMPLE 2: Pattern Searching");
    println!("{}", "=".repeat(70));

    let text = "the quick brown fox jumps over the lazy dog";
    let pattern = "the";

    let sa = SuffixArray::new(text);
    let matches = sa.pattern_search(pattern);

    println!("Text: '{}'", text);
    println!("Pattern: '{}'", pattern);
    println!("Matches at positions: {:?}\n", matches);

    for &pos in &matches {
        println!(
            "  Position {}: '{}'",
            pos,
            &text[pos..pos + pattern.len()]
        );
    }
    println!();
}

fn example_longest_repeated_substring() {
    println!("{}", "=".repeat(70));
    println!("EXAMPLE 3: Longest Repeated Substring");
    println!("{}", "=".repeat(70));

    let text = "abracadabra";
    let sa = SuffixArray::new(text);
    let lrs = sa.longest_repeated_substring();

    println!("Text: '{}'", text);
    println!("Longest repeated substring: '{}'\n", lrs);
}

fn example_count_distinct() {
    println!("{}", "=".repeat(70));
    println!("EXAMPLE 4: Count Distinct Substrings");
    println!("{}", "=".repeat(70));

    let text = "abab";
    let sa = SuffixArray::new(text);
    let count = sa.count_distinct_substrings();

    println!("Text: '{}'", text);
    println!("Number of distinct substrings: {}\n", count);
}

fn example_longest_common_substring() {
    println!("{}", "=".repeat(70));
    println!("EXAMPLE 5: Longest Common Substring");
    println!("{}", "=".repeat(70));

    let text1 = "algorithms";
    let text2 = "altruistic";
    let lcs = longest_common_substring(text1, text2);

    println!("Text 1: '{}'", text1);
    println!("Text 2: '{}'", text2);
    println!("Longest common substring: '{}'\n", lcs);
}

fn example_benchmark() {
    println!("{}", "=".repeat(70));
    println!("EXAMPLE 6: Performance Benchmarking");
    println!("{}", "=".repeat(70));

    // Small text
    let small_text = "banana".repeat(10);
    let small_time = benchmark_construction(&small_text);
    println!(
        "Small text (length {}): {:?}",
        small_text.len(),
        small_time
    );

    // Medium text
    let medium_text = "abracadabra".repeat(100);
    let medium_time = benchmark_construction(&medium_text);
    println!(
        "Medium text (length {}): {:?}\n",
        medium_text.len(),
        medium_time
    );
}

fn main() {
    println!("{}", "=".repeat(70));
    println!("SUFFIX ARRAY AND LCP ARRAY IN RUST");
    println!("{}", "=".repeat(70));
    println!();

    example_basic_construction();
    example_pattern_search();
    example_longest_repeated_substring();
    example_count_distinct();
    example_longest_common_substring();
    example_benchmark();

    println!("{}", "=".repeat(70));
    println!("All examples completed successfully!");
    println!("{}", "=".repeat(70));
}
